packages/licensing/src/class-licensing.php

```diff
@@ -0,0 +1,219 @@
+		}
+	}
+
+	/**
+	 * Get all stored licenses.
+	 *
+	 * @return string[] License keys.
+	 */
+	public function stored_licenses() {
+		$licenses = (array) get_option( self::LICENSES_OPTION_NAME, array() );
+		$licenses = array_filter( $licenses, 'is_scalar' );
+		$licenses = array_map( 'strval', $licenses );
+		$licenses = array_filter( $licenses );
+
+		return $licenses;
+	}
+
+	/**
+	 * Attach the given licenses.
+	 *
+	 * @param string[] $licenses Licenses to attach.
+	 * @return array|WP_Error Results for each license (which may include WP_Error instances) or a WP_Error instance.
+	 */
+	public function attach_licenses( array $licenses ) {
+		if ( ! $this->connection()->is_active() ) {
+			return new WP_Error( 'not_connected', __( 'Jetpack is not connected.', 'jetpack' ) );
+		}
+
+		if ( empty( $licenses ) ) {
+			return array();
+		}
+
+		$xml = $this->request( array( 'user_id' => JETPACK_MASTER_USER ) );
+
+		foreach ( $licenses as $license ) {
+			$xml->addCall( 'jetpack.attachLicense', $license );
+		}
+
+		$xml->query();
+
+		if ( $xml->isError() ) {
+			$error = new WP_Error( 'request_failed', __( 'License attach request failed.', 'jetpack' ) );
+			$error->add( $xml->getErrorCode(), $xml->getErrorMessage() );
+			return $error;
+		}
+
+		$results = array_map(
+			function ( $response ) {
+				if ( isset( $response['faultCode'] ) || isset( $response['faultString'] ) ) {
+					return new WP_Error( $response['faultCode'], $response['faultString'] );
+				}
+
+				return $response;
+			},
+			(array) $xml->getResponse()
+		);
+
+		return $results;
+	}
+
+	/**
+	 * Attach all stored licenses.
+	 *
+	 * @return array|WP_Error Results for each license (which may include WP_Error instances) or a WP_Error instance.
+	 */
+	public function attach_stored_licenses() {
+		$licenses = $this->stored_licenses();
+		$results  = $this->attach_licenses( $licenses );
+
+		if ( is_wp_error( $results ) ) {
+			if ( 'request_failed' === $results->get_error_code() ) {
+				$this->log_error(
+					__( 'Failed to attach your Jetpack license(s). Please try reconnecting Jetpack.', 'jetpack' )
+				);
+			}
+		} else {
+			$failed = array();
+
+			foreach ( $results as $index => $result ) {
+				if ( isset( $licenses[ $index ] ) && is_wp_error( $result ) ) {
+					$failed[] = $licenses[ $index ];
+				}
+			}
+
+			if ( ! empty( $failed ) ) {
+				$this->log_error(
+					sprintf(
+						/* translators: %s is a comma-separated list of license keys. */
+						__( 'The following Jetpack licenses are invalid, already in use or revoked: %s', 'jetpack' ),
+						implode( ', ', $failed )
+					)
+				);
+			}
+		}
+
+		return $results;
+	}
+
+	/**
+	 * Attach all stored licenses during connection flow for the master user.
+	 *
+	 * @return void
+	 */
+	public function attach_stored_licenses_on_connection() {
+		$master_user_id = Jetpack_Options::get_option( 'master_user' );
+		$is_master_user = $master_user_id && get_current_user_id() === $master_user_id;
```

Comment:
I see this code being duplicated throughout Jetpack, even though we have it extracted.
So you could use the `Connection_Manager::is_connection_owner()` instead, to avoid further duplication.

In theory, this code will never be called for non-master users.
Good thinking though, makes a lot of sense to double check it here.